### PR TITLE
enh(web): tpl manage

### PR DESCRIPTION
**Pull request by sc979**
## Description

Please include a short resume of the changes and what is the purpose of PR. Any relevant information should be added to help:
* **QA Team** (Quality Assurance) with tests.
* **reviewers** to understand what are the stakes of the pull request.

**Fixes** # (MON-5007)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [x] 20.04.x (master)

## Checklist

- [x] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/class/centreonHost.class.php [hunk 1]
```diff
@@ -114,6 +117,80 @@ public function getList($enable = false, $template = false, $exclude = null)
...
+            'operatingsystems-windows-snmp'
+        );
+        $ppList = array();
+        $dbResult = $this->db->query('SELECT `name` FROM modules_informations WHERE `name` = "centreon-pp-manager"');
+        if (!$dbResult->fetch() || true === $this->isAllowed()) {
+            return $ppList;
+        }
+        $dbResult = $this->db->query(
+            'SELECT ph.host_id 
+            FROM centreon.mod_ppm_pluginpack_host ph, mod_ppm_pluginpack pp
```
Comment (sc979) [inline www/class/centreonHost.class.php:183]:
```suggestion
            FROM mod_ppm_pluginpack_host ph, mod_ppm_pluginpack pp
```

File: www/class/centreonHost.class.php [hunk 2]
```diff
@@ -1312,6 +1383,34 @@ public function getTemplateChain(
         return $templates;
     }
 
+    /**
+     * @param $hostId
+     * @param $alreadyProcessed
+     * @return array
+     * @throws Exception
+     */
+    public function getHostChain(
```
Comment (callapa) [inline www/class/centreonHost.class.php:1392]: Why return an array if you use its reference to update it?

File: www/class/centreonHost.class.php [hunk 3]
```diff
@@ -1404,6 +1501,44 @@ public function getInheritedValues(
...
+        }
+
+        $licenceManager = $container[\CentreonLicense\ServiceProvider::LM_LICENSE];
+        if (!$licenceManager->validate()) {
+            return false;
+        }
+        $licenseData = ((int)$licenceManager->getData()['licensing']['hosts']) ?? 0;
+        $num = $this->getHostNumber();
+
+        return ($licenseData === -1 || $licenseData >= $num);
```
Comment (callapa) [inline www/class/centreonHost.class.php:1539]:
```suggestion
        return ($licenseData === -1 || $licenseData > $num);
```

File: www/class/centreonHost.class.php (hunk 2 above)
Comment (sc979) [reply www/class/centreonHost.class.php:1392]: to get the data from the n-th level

File: www/class/centreonHost.class.php [hunk 4]
```diff
@@ -114,6 +117,80 @@ public function getList($enable = false, $template = false, $exclude = null)
...
+    }
+
+    /**
+     *  get number of hosts
+     *
+     * @return int
+     */
+    private function getHostNumber(): int
+    {
+        $query = $this->db->query('SELECT COUNT(*) AS `num` FROM host WHERE host_register <> "0"');
```
Comment (callapa) [inline www/class/centreonHost.class.php:152]: It is necessary to be stricter, select directly the right type
```suggestion
        $query = $this->db->query('SELECT COUNT(*) AS `num` FROM host WHERE host_register = "1"');
```

File: www/class/centreonHost.class.php [hunk 5]
```diff
@@ -114,6 +117,80 @@ public function getList($enable = false, $template = false, $exclude = null)
...
+     * get the template currently saved for this host
+     *
+     * @param $hostId
+     *
+     * @return array
+     */
+    public function getSavedTpl($hostId): array
+    {
+        $mTp = [];
+        $dbResult = $this->db->prepare(
```
Comment (callapa) [inline www/class/centreonHost.class.php:130]: Joining with the host table is not necessary.

File: www/class/centreonHost.class.php [hunk 6]
```diff
@@ -1312,6 +1383,34 @@ public function getTemplateChain(
...
+            $stmt->bindParam(':hostId', $hostId, PDO::PARAM_INT);
+            $dbResult = $stmt->execute();
+            if (!$dbResult) {
+                throw new \Exception("An error occured");
+            }
+            while ($row = $stmt->fetch()) {
+                $this->getHostChain($row['host_host_id'], $alreadyProcessed);
+            }
+        }
+        return $alreadyProcessed;
```
Comment (callapa) [inline www/class/centreonHost.class.php:1411]: No need to return the variable. It is passed by reference to the method.

File: www/class/centreonHost.class.php (hunk 4 above)
Comment (sc979) [reply www/class/centreonHost.class.php:152]: shouldn't we consider the AD hosts too ?

File: www/class/centreonHost.class.php (hunk 5 above)
Comment (sc979) [reply www/class/centreonHost.class.php:130]: it is required to get the templates name displayed in the select2

File: www/class/centreonHost.class.php [hunk 7]
```diff
@@ -1312,6 +1383,33 @@ public function getTemplateChain(
         return $templates;
     }
 
+    /**
+     * @param $hostId
+     * @param $alreadyProcessed
+     * @return array
```
Comment (callapa) [inline www/class/centreonHost.class.php:1389]:
```suggestion
```

File: www/class/centreonHost.class.php [hunk 8]
```diff
@@ -1312,6 +1383,32 @@ public function getTemplateChain(
...
+     * @throws Exception
+     */
+    private function getHostChain(
+        $hostId,
+        &$alreadyProcessed
+    ) {
+        if (!in_array($hostId, $alreadyProcessed)) {
+            $alreadyProcessed[$hostId] = $hostId;
+            $query = 'SELECT host_host_id FROM host_template_relation htr 
+            WHERE htr.host_tpl_id = :hostId
```
Comment (sc979) [inline www/class/centreonHost.class.php:1398]:
```suggestion
                WHERE htr.host_tpl_id = :hostId 
```

File: www/class/centreonHost.class.php [hunk 9]
```diff
@@ -1312,6 +1383,32 @@ public function getTemplateChain(
...
+     */
+    private function getHostChain(
+        $hostId,
+        &$alreadyProcessed
+    ) {
+        if (!in_array($hostId, $alreadyProcessed)) {
+            $alreadyProcessed[$hostId] = $hostId;
+            $query = 'SELECT host_host_id FROM host_template_relation htr 
+            WHERE htr.host_tpl_id = :hostId 
+            ORDER BY `order` ASC';
```
Comment (sc979) [inline www/class/centreonHost.class.php:1399]:
```suggestion
                WHERE htr.host_tpl_id = :hostId 
                ORDER BY `order` ASC';
```

**Pull request merged**